### ISSUE TITLE
capture_dynamic_output_shape_ops=True changing expected output between eager and compiled versions

### ISSUE THREAD

**Issue by krzysztofjordan** [pytorch/pytorch#130290]
### 🐛 Describe the bug

There seems to be a difference in what eager mode and compiled mode outputs for the code snippet that follows. 

TL;DR the code is trying to take two tensors and map the elements of the first tensor to the indices of the second tensors where the values are equal or 0 if the element does not exist in the second tensor.

Below is a minimal repro for the issue:

```python
import torch

@torch.compile
@torch._dynamo.config.patch(capture_dynamic_output_shape_ops=True)
def minimal_repro(
    tensor: torch.Tensor,
    mapping: torch.Tensor,
) -> torch.Tensor:
    xx, yy = torch.meshgrid(mapping, tensor, indexing="ij")
    indices = torch.argwhere(xx == yy)

    mapped_values = torch.zeros_like(tensor)
    mapped_values[indices[:, 1]] = indices[:, 0]

    return mapped_values

def test_minimal_repro() -> None:
    tensor = torch.tensor([1, 2, 3, 5, 6, 7])
    mapping = torch.tensor([0, 3, 4, 5, 7])

    mapped_tensor = minimal_repro(tensor, mapping)
    torch.testing.assert_close(mapped_tensor, torch.tensor([0, 0, 1, 3, 0, 4]))
```

The test_minimal_repro code works with 
```python
 @torch._dynamo.config.patch(capture_dynamic_output_shape_ops=True)
``` 
being commented out but fails when the annotation is enabled.
mapped_tensor has the following value tensor([0, 0, 4, 1, 3, 0])

### Versions

main branch

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire

**Comment by ZhaoqiongZ** [comment]
Hi @krzysztofjordan , I can not reproduce this issue with both nightly version  2.5.0.dev20240708+cu118 and compiled version with torch==2.5.0a0+git6c4efd4
Could you please specify the specific commit and how you compile in main branch?

**Comment by soulitzer** [comment]
I was able to repro on an older commit e6d4451ae8987bf8d6ad85eb7cde685fac746f6f (6/18) and only on inductor, maybe it has been fixed in a later version.

**Comment by soulitzer** [comment]
Repro in issue runs with cpu, but the issue also exists when tensors are on cuda

**Comment by soulitzer** [comment]
argwhere aka nonzero returns a non-contiguous tensor with shape (3, 2), stride (1, 3), but the triton kernel that applies index_put seems to assume that the input is contiguous, last stride is 1.
```python
def triton_(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (1 + x0), xmask)
    tmp6 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.full([XBLOCK], 6, tl.int32)
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    tl.device_assert(((0 <= tmp4) & (tmp4 < 6)) | ~(xmask), "index out of bounds: 0 <= tmp4 < 6")
    tl.store(out_ptr0 + (tl.broadcast_to(tmp4, [XBLOCK])), tmp6, xmask)
 ```
 
When `tmp0 = tl.load(in_ptr0 + (1 + x0), xmask)` is changed to `3` the result is correct.

**Comment by ezyang** [comment]
This sounds like the meta for argwhere is wrong (not producing correct strides)

**Comment by zou3519** [comment]
@soulitzer are you still working on this?

**Comment by soulitzer** [comment]
Not actively, the latest status is here https://github.com/pytorch/pytorch/pull/130440#issuecomment-2263334177, someone else can pick up if they'd like.

**Comment by dberardi99** [comment]
I could have a try if it is still open

**Comment by soulitzer** [comment]
Feel free to submit a PR @dberardi99

**Comment by Chillee** [comment]
I will ask Jeffrey about the update on this.

**Comment by eellison** [comment]
cc @bobrenjc93 any interest in this ?

**Comment by bobrenjc93** [comment]
looking into this now...

in eager mode

```
>>> indices.stride()
(1, 3)
```

but in tlparse of dynamic version we have strides of `(2, 1)`

```
indices: "i64[u0, 2][2, 1]cpu" = torch.argwhere(eq);  eq = None
```

that seems sketchy

**Comment by bobrenjc93** [comment]
https://github.com/pytorch/pytorch/pull/145650 should fix it

**Comment by bobrenjc93** [comment]
Actually I just realized @ezyang is doing the same thing but farther along in the CI whack-a-mole https://github.com/pytorch/pytorch/pull/144695. Maybe it makes sense to patch on the following test case to that PR?

```
    @torch._dynamo.config.patch(capture_dynamic_output_shape_ops=True)
    def test_argwhere_with_dynamic_shapes(self):
        def fn(
            tensor: torch.Tensor,
            mapping: torch.Tensor,
        ) -> torch.Tensor:
            xx, yy = torch.meshgrid(mapping, tensor, indexing="ij")
            indices = torch.argwhere(xx == yy)

            mapped_values = torch.zeros_like(tensor)
            mapped_values[indices[:, 1]] = indices[:, 0]

            return mapped_values

        tensor = torch.tensor([1, 2, 3, 5, 6, 7])
        mapping = torch.tensor([0, 3, 4, 5, 7])
        opt = torch.compile(fn, fullgraph=True)
        self.assertEqual(fn(tensor, mapping), opt(tensor, mapping))
```

**Comment by ezyang** [comment]
sure, feel free to toss that test into its own pr
